### PR TITLE
Add vectorscan (same abi as hyperscan) for arm 64

### LINKED ISSUE #29276: Add Vectorscan (same ABI as Hyperscan) for Arm 64
*Title*: *Add Hyperscan/Vectorscan to non x86 processors like Arm 64*

*Description*:
https://github.com/envoyproxy/envoy/issues/19376 added hyperscan as a matcher and an optional regular expression engine. 
Hyperscan supports x86 processors and is included only in contib build for intel.

Vectorscan - https://github.com/VectorCamp/vectorscan is a fork of Hyperscan supporting the same ABI for Arm64 and PowerPC 64. Specificaly:
- Arm 64 - ARM NEON/ASIMD and/or  ARM SVE2
- PowerPC 64 - VSX

Either switch to Vectorscan for all supported processors (including Intel), or use Vectorscan for supported non intel processors.

>Describe the desired behavior, what scenario it enables and how it
would be used.
The Hyperscan functionality will be available for all Envoy supported CPU architectures.

[optional *Relevant Links*:]
https://github.com/VectorCamp/vectorscan 

>Any extra documentation required to understand the issue.

### PR DESCRIPTION
**Pull request by sambercovici**
Commit Message: Fix #29276 
Additional Description: Use Vectorscan for Arm 64 CPUs
Risk Level: Low
Testing: Covered by original commit
Docs Changes:
Release Notes:
Platform Specific Features: aarch64

### CODE REVIEW

**Comment by sambercovici** [comment]
@phlax , can you please assist and let me know why the https://dev.azure.com/cncf/envoy/_build/results?buildId=150625&view=results failed?

File: bazel/foreign_cc/vectorscan.patch
Comment (htuch) [inline bazel/foreign_cc/vectorscan.patch:9]: Can you fix this upstream?

File: bazel/repository_locations.bzl
Comment (htuch) [inline bazel/repository_locations.bzl:488]: Can you run through https://github.com/envoyproxy/envoy/blob/main/DEPENDENCY_POLICY.md?

Given regex processing can in many paces be considered security sensitive (correctness of rule matching, CPU/memory DoS for example), I'd like to weigh up the win we're getting from this dependency, which seems to be a community fork.

Comment (sambercovici) [reply bazel/repository_locations.bzl:488]: Hi @htuch 

The vectorscan is a company (https://www.vectorcamp.gr/) and community led fork (under paid contract: https://www.vectorcamp.gr/2020/05/09/we-are-thrilled-to-announce-our-first-customer/) to support hyperscan ABI in Arm CPUs and additional CPUs with the same license, same tests, scans and fixes. 
see: https://github.com/envoyproxy/envoy/issues/29276#issuecomment-1709620894 for additional information on CI builds, tests, etc.

Let me know if you want me to fill the items as described in: https://github.com/envoyproxy/envoy/blob/main/DEPENDENCY_POLICY.md

-Sam.

Comment (htuch) [reply bazel/repository_locations.bzl:488]: Thanks - yeah, please fill in the table so we can evaluate this new dep.

File: bazel/foreign_cc/vectorscan.patch
Comment (sambercovici) [reply bazel/foreign_cc/vectorscan.patch:9]: opened https://github.com/VectorCamp/vectorscan/issues/176 waiting for response on ISSUE-1 there

File: bazel/repository_locations.bzl
Comment (sambercovici) [reply bazel/repository_locations.bzl:488]: Adding https://github.com/VectorCamp/vectorscan/issues/177 as reference.

- Cloud Native Computing Foundation (CNCF) [approved license](https://github.com/cncf/foundation/blob/master/allowed-third-party-license-policy.md#approved-licenses-for-allowlist) - BSD License - https://github.com/VectorCamp/vectorscan/blob/vectorscan/5.4.10.1/LICENSE

- Dependencies must not substantially increase the binary size unless they are optional (i.e. confined to specific extensions) - same as hyperscan which is approved.

- No duplication of existing dependencies - Hyperscan does not cover Arm CPUs, Vectorscan is used instead.

- Hosted on a git repository and the archive fetch must directly reference this repository. We will NOT support intermediate artifacts built by-hand located on GCS, S3, etc. - YES - https://github.com/VectorCamp/vectorscan

- CVE history appears reasonable, no pathological CVE arcs - Should be similar to Hyperscan

- Code review (ideally PRs) before merge - YES - see: https://github.com/envoyproxy/envoy/issues/29276#issuecomment-1709620894

- Security vulnerability process exists, with contact details and reporting/disclosure process - YES - see: https://github.com/VectorCamp/vectorscan/issues/177#issuecomment-1743037604

- > 1 contributor responsible for a non-trivial number of commits - YES - see: https://github.com/VectorCamp/vectorscan/graphs/contributors

- Tests run in CI - YES - see: https://github.com/envoyproxy/envoy/issues/29276#issuecomment-1709620894 and https://github.com/VectorCamp/vectorscan/issues/177#issuecomment-1743037604

- High test coverage (also static/dynamic analysis, fuzzing) - YES - see: https://github.com/envoyproxy/envoy/issues/29276#issuecomment-1709620894

- Envoy can obtain advanced notification of vulnerabilities or of security releases - Not sure

- Do other significant projects have shared fate by using this dependency? - [ClickHouse ](https://github.com/ClickHouse)

- Releases (with release notes) - YES 

- Commits/releases in last 90 days - YES - see: https://github.com/VectorCamp/vectorscan/graphs/commit-activity

Comment (moderation) [reply bazel/repository_locations.bzl:488]: OSSF Scorecard result is low. Security policy and code review would help.
```
scorecard --repo=https://github.com/VectorCamp/vectorscan/
RESULTS
-------
Aggregate score: 4.1 / 10

Check scores:
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
|  SCORE  |          NAME          |             REASON             |                                               DOCUMENTATION/REMEDIATION                                               |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Binary-Artifacts       | no binaries found in the repo  | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#binary-artifacts       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 1 / 10  | Branch-Protection      | branch protection is not       | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#branch-protection      |
|         |                        | maximal on development and all |                                                                                                                       |
|         |                        | release branches               |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | CI-Tests               | 0 out of 6 merged PRs          | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#ci-tests               |
|         |                        | checked by a CI test -- score  |                                                                                                                       |
|         |                        | normalized to 0                |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | CII-Best-Practices     | no effort to earn an OpenSSF   | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#cii-best-practices     |
|         |                        | best practices badge detected  |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 1 / 10  | Code-Review            | found 6 unreviewed changesets  | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#code-review            |
|         |                        | out of 7 -- score normalized   |                                                                                                                       |
|         |                        | to 1                           |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Contributors           | 6 different organizations      | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#contributors           |
|         |                        | found -- score normalized to   |                                                                                                                       |
|         |                        | 10                             |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Dangerous-Workflow     | no workflows found             | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#dangerous-workflow     |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Dependency-Update-Tool | no update tool detected        | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#dependency-update-tool |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Fuzzing                | project is not fuzzed          | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#fuzzing                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 9 / 10  | License                | license file detected          | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#license                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Maintained             | 30 commit(s) out of 30 and 0   | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#maintained             |
|         |                        | issue activity out of 30 found |                                                                                                                       |
|         |                        | in the last 90 days -- score   |                                                                                                                       |
|         |                        | normalized to 10               |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Packaging              | no published package detected  | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#packaging              |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Pinned-Dependencies    | no dependencies found          | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#pinned-dependencies    |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | SAST                   | SAST tool is not run on all    | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#sast                   |
|         |                        | commits -- score normalized to |                                                                                                                       |
|         |                        | 0                              |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Security-Policy        | security policy file not       | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#security-policy        |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Signed-Releases        | no releases found              | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#signed-releases        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Token-Permissions      | no github tokens found         | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#token-permissions      |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Vulnerabilities        | no vulnerabilities detected    | https://github.com/ossf/scorecard/blob/7a94273ba14c1c702688f9efb57754d9fe112519/docs/checks.md#vulnerabilities        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
```

Comment (markos) [reply bazel/repository_locations.bzl:488]: fwiw, all PRs are CI tested (about 100 configurations), unfortunately there is a new CI in development now (buildbot-based) which is not yet integrated with Github.

Comment (markos) [reply bazel/repository_locations.bzl:488]: Also, we do the Debian packaging ourselves, it's directly uploaded to Debian and Ubuntu.

Comment (sambercovici) [reply bazel/repository_locations.bzl:488]: The score is artificially low. see: https://github.com/envoyproxy/envoy/issues/29276#issuecomment-1709620894

File: bazel/foreign_cc/vectorscan.patch
Comment (sambercovici) [reply bazel/foreign_cc/vectorscan.patch:9]: https://github.com/VectorCamp/vectorscan/issues/180

**Comment by sambercovici** [comment]
Hi @htuch , how can we progress?

**Comment by htuch** [comment]
I think that given we're in the same position as Hyperscan, this is `contrib/` only anyway and most tablestakes from our policy are there, we should go ahead.

File: bazel/foreign_cc/vectorscan.patch
Comment (htuch) [reply bazel/foreign_cc/vectorscan.patch:9]: Will this be fixed before we merge?

Comment (markos) [reply bazel/foreign_cc/vectorscan.patch:9]: You may want to wait a few days, it will be included yes, but I'm fixing clang 15/16 issues right now as they are higher priority and ironing out a few issues with the new CI. I'll post an update here when that happens.

**Comment by jmarantz** [comment]
@htuch what's the next step; are more changes required or should we add more review comments? Or is it time to close this.

I have no horse in this race and have not read the code; just trying to figure out how triage the PR as maintainer oncall.

**Comment by htuch** [comment]
I was waiting for https://github.com/envoyproxy/envoy/pull/29881/files#r1348331292 to be resolved, CC @markos who has the AI on that.

**Comment by markos** [comment]
@htuch the new release of vectorscan with this fix, will happen the next days. In the meantime, I just learned myself about recent changes in hyperscan license: https://networkbuilders.intel.com/docs/networkbuilders/accelerate-snort-performance-with-hyperscan-and-intel-xeon-processors-on-public-clouds-1680176363.pdf (section 2.2)
In essence, versions 5.5 onwards are closed/proprietary license.

You might want to consider that for the future.

**Comment by htuch** [comment]
Relevant passage is:

> The latest open-source version (BSD-3 license) of Hyperscan on Github is 5.4. Intel conducts continuous internal
development and delivers new Hyperscan releases under Intel Proprietary License (IPL) beginning from 5.5 for interested
customers. Please contact authors to learn more about getting new Hyperscan releases.

I think this means we will be unable to promote Hyperscan to a non-contrib extension, unless there is some LTS assurance on security releases. What will Vectorscan do going forward?

CC @envoyproxy/dependency-shepherds

**Comment by markos** [comment]
@htuch we intend to continue development entirely as there is a lot of demand esp on Arm, ofc we intend to keep Intel working, add more platforms in the near future (Power is already there, Loongson is almost working, plus we have a C backend ready to merge) unfortunately this means we can no longer track API changes from later versions of hyperscan, unless it be with reverse engineering, but we would rather not take that path as it would put us in a difficult situation against Intel. But this is also a good thing for us as it removes the obstacles we have for future changes and advances.

**Comment by sambercovici** [comment]
@htuch, does it make sense to switch to use Vectorscan for all target CPUs (including Intel) if Hyperscan does not provide any LTS for 5.4?

**Comment by zhxie** [comment]
Hyperscan is still fixing bugs and any security concern issues in v5.4, the latest version of Hyperscan is v5.4.2, which has been released in April. See https://github.com/intel/hyperscan/releases.

**Comment by htuch** [comment]
> Hyperscan is still fixing bugs and any security concern issues in v5.4, the latest version of Hyperscan is v5.4.2, which has been released in April. See https://github.com/intel/hyperscan/releases.
> 
> And the dependency bot is still tracking its update. See #26849.

@zhxie do you know how long security updates will continue?

**Comment by zhxie** [comment]
AFAIK, Hyperscan 5.4 is still actively maintained. It has a valid CPE `cpe:2.3:a:intel:hyperscan:*` and we are still tracking on its vulnerabilities. All bugfixes and security-related patches in 5.5+ will be backported to 5.4, while performance-related (e.g., novel algorithms) will not be made public.

**Comment by htuch** [comment]
Yep, I agree it's actively maintained, but are you aware of any policy around how long this will be for? Knowledge of this is consequential right now, since if we're not going to see LTS and Vectorscan offers a viable replacement, it seems worth taking seriously.

**Comment by markos** [comment]
> Thanks @zhxie. @markos do you plan on porting all the bug fixes and security patches from Hyperscan 5.4 to Vectorscan?

This has been a commitment from us since the beginning of the project.

**Comment by mattklein123** [comment]
Friendly ping @htuch what's the plan here?

/wait-any

**Comment by htuch** [comment]
Are we at the point we can get rid of the patch? If so, let's do that and merge. If it's going to be a ton of time until we can get rid of the patch, merge with an issue tracking the need to get rid of it. If it's some small bounded time until upstream is caught up, let's wait and then merge without the patch.

**Comment by markos** [comment]
> Are we at the point we can get rid of the patch? If so, let's do that and merge. If it's going to be a ton of time until we can get rid of the patch, merge with an issue tracking the need to get rid of it. If it's some small bounded time until upstream is caught up, let's wait and then merge without the patch.

Hi, the patch is already merged in develop branch, but I cannot do a release yet, I have yet to fix this https://github.com/VectorCamp/vectorscan/issues/184 first.

**Comment by htuch** [comment]
@markos do you have an ETA?

**Comment by markos** [comment]
> @markos do you have an ETA?

When https://github.com/VectorCamp/vectorscan/issues/184 is done, sorry I cannot release until that is fixed.

**Comment by markos** [comment]
> > @markos do you have an ETA?
> 
> When [VectorCamp/vectorscan#184](https://github.com/VectorCamp/vectorscan/issues/184) is done, sorry I cannot release until that is fixed.

This issue is now fixed, I will do an new release today or over the weekend. Thanks for being patient.

**Comment by markos** [comment]
@htuch 5.4.11 has just been released: https://github.com/VectorCamp/vectorscan/releases/tag/vectorscan%2F5.4.11

**Comment by sambercovici** [comment]
@htuch , @phlax I updated the PR to use the latest Vectorscan (5.4.11) and removed the patch.
Please let me know if something else is missing.

**Comment by markos** [comment]
@sambercovici @htuch something was left out in the installation, turns out that testing 100 configurations in the CI is not enough if you forget to test the installation process. Apologies for that, will do an immediate fix release asap.

**Review by htuch — APPROVED** [review]

**Comment by htuch** [comment]
@markos LMK when fixed and I can merge.

/wait

**Comment by markos** [comment]
@htuch https://github.com/VectorCamp/vectorscan/releases/tag/vectorscan%2F5.4.11 thanks for being patient.

**Comment by sambercovici** [comment]
updated sha256 to use the latest 5.4.11 tar.gz.
@htuch , @markos , let me know if anything else is needed.
Otherwise, please approve and merge.

**Comment by markos** [comment]
@sambercovici @htuch I've already uploaded it to Debian: https://buildd.debian.org/status/package.php?p=vectorscan nothing else is left on my side. Thanks for supporting the project.

**Comment by sambercovici** [comment]
Hi @htuch ,

@markos added for vectorscan 5.4.11 mandatory dependencies on pkg-config and libsqlite3-dev.
The original hyperscan does not mandate pgk-config and sqlite3.

Envoy dev/build environment does not include those dependency. 
I don't feel comfortable adding those dependencies to address the changes in 5.4.11.
Instead, I created a patch that 
(a) reverts https://github.com/VectorCamp/vectorscan/pull/188 and 
(b) adds a flag to disable the detection of sqlite3 altogether.

@markos , please consider making pkg-config optional and detect that sqlite3 does not exists, so I can remove that patches.
@htuch, please let me know if you want to handle this differently, or otherwise please approve.

-Sam.

**Comment by markos** [comment]
> @markos , please consider making pkg-config optional and detect that sqlite3 does not exists, so I can remove that patches. @htuch, please let me know if you want to handle this differently, or otherwise please approve.

I have no objections making sqlite optional, but pkg-config will stay. I find no reason to duplicate its functionality in the cmakefiles and in the future other possible dependencies will also be able to use it. It's in every distro and it's a really small dependency.

**Review by htuch — APPROVED** [review]
LGTM but please followup to remove this patch!

**Comment by htuch** [comment]
OK, this has lingered for super long so let's jut merge but I think we shouldn't need an sqlite dependency and I'm hoping there is some clean way to resolve the pkgconfig thing.

**Pull request merged**
